_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+
+If the RNN in question is a GRU (and assuming that of the outputs, we take just the final time step, which is what we've been doing throughout), there really is no difference: The final state equals the final output. If it's an LSTM, however, there is a second, the "cell", state. In that case, returning the state instead of the final output will carry more information.
+
+```{r}
+encoder_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, num_layers = 1, dropout = 0) {
+    
+    self$type <- type
+    
+    self$rnn <- if (self$type == "gru") {
+      nn_gru(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    } else {
+      nn_lstm(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    }
+    
+  },
+  
+  forward = function(x) {
+    
+    x <- self$rnn(x)
+    
+    # return last states for all layers
+    # per layer, a single tensor for GRU, a list of 2 tensors for LSTM
+    x <- x[[2]]
+    x
+    
+  }
+  
+)
+```
+
+### Decoder
+
+In the decoder, too, the main component is an RNN. In contrast to previously-shown architectures, though, it does not just return a prediction. It also reports back the RNN's final state.
+
+```{r}
+decoder_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, num_layers = 1) {
+    
+    self$type <- type
+    
+    self$rnn <- if (self$type == "gru") {
+      nn_gru(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        batch_first = TRUE
+      )
+    } else {
+      nn_lstm(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        batch_first = TRUE
+      )
+    }
+    
+    self$linear <- nn_linear(hidden_size, 1)
+    
+  },
+  
+  forward = function(x, state) {
+    
+    # input to forward:
+    # x is (batch_size, 1, 1)
+    # state is (1, batch_size, hidden_size)
+    x <- self$rnn(x, state)
+    
+    # break up RNN return values
+    # output is (batch_size, 1, hidden_size)
+    # next_hidden is
+    c(output, next_hidden) %<-% x
+    
+    output <- output$squeeze(2)
+    output <- self$linear(output)
+    
+    list(output, next_hidden)
+    
+  }
+  
+)
+```
+
+### `seq2seq` module
+
+In `seq2seq` is where the action happens. The plan is to encode once, then call the decoder in a loop.
+
+If you look back to decoder `forward()`, you see that it takes two arguments: `x` and `state`.
+
+Dependent on the context, `x` corresponds to one of three things.
```

Comment:
```suggestion
Depending on the context, `x` corresponds to one of three things.
```

Reply:
What are the three things? If they're listed below, it's a bit unclear what the exact correspondence is, so you might give abbreviated versions of them here.

Reply:
oh, yes, I see... will mention them here already

Reply:
@batpigandme if by chance you see this - could you let me know if this is correct, especially, if this is a correct use of "leading"? I found "leading" as a synonym for "preceding", but have no idea if it can be used like that (actually, that sense of "leading" I was only familiar with from window function arguments - lag & lead :-))

> Depending on the context, `x` corresponds to one of three things: final input, preceding prediction, or leading ground truth.

Reply:
For leading do you mean prior? I'm not 100% clear on what its intended meaning is here. 😬

Reply:
yes, same as preceding/previous/antecedent/....?

but now that you're suggesting it: **prior** might be best :-)